pkg/csi-handler/pv_checker.go

```diff
@@ -203,6 +219,46 @@ func (checker *PVHealthConditionChecker) CheckNodeVolumeStatus(kubeletRootPath s
 
 	if volumeCondition.GetAbnormal() {
 		checker.eventRecorder.Event(pod, v1.EventTypeWarning, "VolumeConditionAbnormal", volumeCondition.GetMessage())
+	} else {
+		return checker.sendRecoveryEventToPod(pod, volumeCondition)
+	}
+
+	return nil
+}
+
+// sendRecoveryEventToPVC sends the recovery event to the pvc
+// If the volume condition is normal and abnormal event wasn't expired,
+// PVHealthConditionChecker should send recovery event.
+func (checker *PVHealthConditionChecker) sendRecoveryEventToPVC(pvc *v1.PersistentVolumeClaim, volumeCondition *VolumeConditionResult) error {
+	pvcUID := string(pvc.ObjectMeta.GetUID())
+	key := fmt.Sprintf("%s:%s:%s", pvcUID, v1.EventTypeWarning, "VolumeConditionAbnormal")
+	events, err := checker.eventInformer.Informer().GetIndexer().ByIndex(util.DefaultEventIndexerName, key)
+	if err != nil {
+		klog.Warningf("Get abnormal event from indexer failed: %+v", err)
+		return nil
+	}
+
+	if len(events) > 0 {
+		checker.eventRecorder.Event(pvc, v1.EventTypeNormal, "VolumeConditionNormal", volumeCondition.GetMessage())
```

Comment:
Maybe we need to generate specific message to indicate this is a recovery event.

Reply:
You mean the message part. Right?
An event usually includes

- type
- reason
- message

Is it ok to set `VolumeConditionNormal` as the reason for the recovery event?

Reply:
Yes, and i am ok with `VolumeConditionNormal`